lib/addressable/template.rb

```diff
@@ -1038,7 +1038,18 @@ def parse_template_pattern(pattern, processor=nil)
 
       # Ensure that the regular expression matches the whole URI.
       regexp_string = "^#{regexp_string}$"
-      return expansions, Regexp.new(regexp_string)
+
```

Comment:
It's been a long time since I looked at this code, but I think the general idea is great! Could probably optimize further while we're at it though.

https://github.com/sporkmonger/addressable/pull/389/files#diff-76daf569c1d16697395cf4be56da5592R984-R987

I feel like we could probably get even more performance wins if we put an early return up towards the top of this function instead - as long as pattern (which is actually immutable, if I remember correctly) and processor (usually nil) passed in matches, we could return a cached version of the expansions and Regexp. Putting the cache at the very bottom of the function reduces Regexp allocations but still lets a lot of cacheable work (which is done by a bunch of String allocations) be done first to compare to the previously cached value. 

One way to approach it might be to rename this old function `parse_fresh_template_pattern` and then make a new `parse_template_pattern` that calls to it only if value isn't already cached for pattern / processor combo. Given processors could be non-deterministic, I might even say only cache if processor is nil?